synapse_data_warehouse/synapse_aggregate/dynamic_tables/V2.55.0__unique_user_downloads_aggregate.sql

```diff
@@ -0,0 +1,211 @@
+                AND project_id IS NULL
+            )
+        GROUP BY
+            ROLLUP(agg_year, agg_quarter, agg_month, agg_day),
+            GROUPING SETS (
+                (),                                                 -- 1) aggregates across every object in Synapse
+                (project_id),                                       -- 2) aggregates across every object in this project
+                (association_object_type),                          -- 3) aggregates across every object of this type in Synapse
+                (project_id, association_object_type),              -- 4) aggregates across every object of this type in this project
+                -- (association_object_type, association_object_id),   -- 5) aggregates across projects for this object
+                (project_id, association_object_type, association_object_id)  -- 6) aggregates for a specific object in a specific project
+            )
+    ),
+    -- Step 2: Determine the aggregate period (grain) and start and end dates
+    agg_period_calculations AS (
+
+        SELECT
+            agg_year,
+            agg_quarter,
+            agg_month,
+            agg_day,
+
+            -- Create `agg_period` column
+            -- Determine granularity based on which dimensions were rolled up
+            CASE
+                WHEN g_year = 1 AND g_quarter = 1 AND g_month = 1 AND g_day = 1 THEN 'ALL TIME'
+                WHEN g_year = 0 AND g_quarter = 1 AND g_month = 1 AND g_day = 1 THEN 'YEARLY'
+                WHEN g_year = 0 AND g_quarter = 0 AND g_month = 1 AND g_day = 1 THEN 'QUARTERLY'
+                WHEN g_year = 0 AND g_quarter = 0 AND g_month = 0 AND g_day = 1 THEN 'MONTHLY'
+                WHEN g_day = 0 THEN 'DAILY'
+            END AS agg_period,
+
+            -- Create `agg_level` column
+            -- Determine granularity based on which grouping sets were rolled up
+            CASE
+                WHEN project_id IS NOT NULL AND association_object_type IS NOT NULL AND association_object_id IS NOT NULL
+                    THEN 'OBJECT WITHIN PROJECT'
+                WHEN project_id IS NOT NULL AND association_object_type IS NOT NULL AND association_object_id IS NULL
+                    THEN 'OBJECT TYPE WITHIN PROJECT'
+                WHEN project_id IS NOT NULL AND association_object_type IS NULL
+                    THEN 'PROJECT'
+                -- This covers other object types, but doesn't match table/file aggregates across projects.
+                -- The check against table/file aggregates is not relevant, since we don't currently include
+                -- the (association_object_type, association_object_id) grouping set, but no harm in including it
+                -- to reduce maintenence burden.
+                WHEN project_id IS NULL AND association_object_type IS NOT NULL AND association_object_id IS NOT NULL AND association_object_type NOT IN ('TableEntity', 'FileEntity')
```

Comment:
I see 5 GROUPING SETS and 6 CASEs for `agg_level`. This one looks like it corresponds to GROUPING SET `(association_object_type, association_object_id)`. Remove?